File: Applets_Container/pac/src/main/java/com/pumpkin/pac/widget/loading/LoadingView.java
package com.pumpkin.pac.widget.loading;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;

import androidx.appcompat.widget.AppCompatImageView;

public class LoadingView extends AppCompatImageView {
    private LoadingDrawable mLoadingDrawable;

    public LoadingView(Context context) {
        super(context);
    }

    public LoadingView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    private void init(Context context) {
        setLoadingRenderer(new ElectricFanLoadingRenderer(context));
    }

    public void setLoadingRenderer(LoadingRenderer loadingRenderer) {
        mLoadingDrawable = new LoadingDrawable(loadingRenderer);
        setImageDrawable(mLoadingDrawable);
    }

//    @Override
//    protected void onAttachedToWindow() {
//        super.onAttachedToWindow();
//        startAnimation();
//    }
//
//    @Override
//    protected void onDetachedFromWindow() {
//        super.onDetachedFromWindow();
//        stopAnimation();
//    }
//
//    @Override
//    protected void onVisibilityChanged(View changedView, int visibility) {
//        super.onVisibilityChanged(changedView, visibility);
//
//        final boolean visible = visibility == VISIBLE && getVisibility() == VISIBLE;
//        if (visible) {
//            startAnimation();
//        } else {
//            stopAnimation();
//        }
//    }

    public void progress(int progress) {
        mLoadingDrawable.change(progress);
        invalidate();
    }
}

File: Applets_Container/base/mvvm/src/main/java/com/pumpkin/mvvm/widget/loading/LoadingView.kt
package com.pumpkin.mvvm.widget.loading

import android.content.Context
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.view.View
import android.widget.ProgressBar
import androidx.core.content.ContextCompat
import com.pumpkin.mvvm.R


class LoadingView
@JvmOverloads
constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = R.attr.progressBarStyle)
    : ProgressBar(context, attrs, defStyleAttr) {
    private var mSprite: Sprite? = null

    init {
        init()
    }

    private fun init() {
        val sprite: Sprite = ThreeBounce()
        sprite.color = ContextCompat.getColor(context, R.color.seed)
        setIndeterminateDrawable(sprite)
    }

    override fun setIndeterminateDrawable(d: Drawable?) {
        if (d !is Sprite) {
            throw IllegalArgumentException("this d must be instance of Sprite");
        }
        setIndeterminateDrawable(d as Sprite?)
    }

    fun setIndeterminateDrawable(d: Sprite?) {
        super.setIndeterminateDrawable(d)
        mSprite = d
        onSizeChanged(width, height, width, height)
        if (visibility == VISIBLE) {
            mSprite?.start()
        }
    }

    override fun getIndeterminateDrawable(): Sprite? {
        return mSprite
    }

    override fun unscheduleDrawable(who: Drawable) {
        super.unscheduleDrawable(who)
        if (who is Sprite) {
            who.stop()
        }
    }

    override fun onWindowFocusChanged(hasWindowFocus: Boolean) {
        super.onWindowFocusChanged(hasWindowFocus)
        if (hasWindowFocus) {
            if (mSprite != null && visibility == VISIBLE) {
                mSprite?.start()
            }
        }
    }

    fun setColor(color: Int) {
        if (mSprite != null) {
            mSprite!!.color = color
        }
        invalidate()
    }


    override fun onScreenStateChanged(screenState: Int) {
        super.onScreenStateChanged(screenState)
        if (screenState == View.SCREEN_STATE_OFF) {
            if (mSprite != null) {
                mSprite?.stop()
            }
        }
    }

    companion object {
        const val TAG = "LoadingView"
    }
}